graphium/trainer/predictor.py

```diff
@@ -449,35 +447,47 @@ def flag_step(self, batch: Dict[str, Tensor], step_name: str, to_cpu: bool) -> D
 
     def on_train_batch_start(self, batch: Any, batch_idx: int) -> Optional[int]:
         self.train_batch_start_time = time.time()
+        self.is_log_train_metrics = (self.metrics_every_n_train_steps is None) or (
+            (batch_idx + 1) % self.metrics_every_n_train_steps != 0
+        )
         return super().on_train_batch_start(batch, batch_idx)
 
     def on_train_batch_end(self, outputs, batch: Any, batch_idx: int) -> None:
-        if (batch_idx + 1) % self.metrics_every_n_steps != 0:
-            return
 
+        # Get the throughput of the batch
         train_batch_time = time.time() - self.train_batch_start_time
         num_graphs = self.get_num_graphs(batch["features"])
         tput = num_graphs / train_batch_time
-        if self.save_preds_and_targets:
-            # this code is likely repeated for validation and testing, this should be moved to a function
-            self.task_epoch_summary.update_predictor_state(
-                step_name="train",
-                targets=outputs["targets"],
-                predictions=outputs["preds"],
-                loss=outputs["loss"],  # This is the weighted loss for now, but change to task-specific loss
-                task_losses=outputs["task_losses"],
-                n_epochs=self.current_epoch,
-            )
-            metrics_logs = self.task_epoch_summary.get_metrics_logs()  # Dict[task, metric_logs]
-            metrics_logs["_global"]["grad_norm"] = self.get_gradient_norm()
-            outputs.update(metrics_logs)  # Dict[task, metric_logs]. Concatenate them?
 
+        # Get the metrics that are logged at every step (loss, grad_norm, batch_time, batch_tput)
         concatenated_metrics_logs = {}  # self.task_epoch_summary.concatenate_metrics_logs(metrics_logs)
         concatenated_metrics_logs["train/loss"] = outputs["loss"]
         outputs["grad_norm"] = self.get_gradient_norm()
         concatenated_metrics_logs["train/grad_norm"] = outputs["grad_norm"]
         concatenated_metrics_logs["train/batch_time"] = train_batch_time
         concatenated_metrics_logs["train/batch_tput"] = tput
+
+        # If logging is skipped for this step, then log the important metrics anyway and return
+        if self.is_log_train_metrics:
+            if self.logger is not None:
+                self.logger.log_metrics(
+                    concatenated_metrics_logs, step=self.global_step
+                )  # This is a pytorch lightning function call
+            return
+
+        # this code is likely repeated for validation and testing, this should be moved to a function
+        self.task_epoch_summary.update_predictor_state(
+            step_name="train",
+            targets=outputs["targets"],
+            predictions=outputs["preds"],
+            loss=outputs["loss"],  # This is the weighted loss for now, but change to task-specific loss
```

Comment:
Also got this error when testing `config_large_gcn.yaml`:
```
`  File "/nethome/zhiyil/graphium/graphium/ipu/ipu_wrapper.py", line 105, in on_train_batch_end
    super().on_train_batch_end(outputs, batch, batch_idx)
  File "/nethome/zhiyil/graphium/graphium/trainer/predictor.py", line 481, in on_train_batch_end
    targets=outputs["targets"],
KeyError: 'targets'
[10:29:23.668] [poptorch::python] [critical] KeyError: 'targets'`
```

Reply:
Yeah I forgot to mention that I didn't test the code at all. It was just easier to do a wick draft than to explain in the other PR what I had in mind. 

If it's not too much to ask, can you double check everything before merging?

Reply:
I tested with debug config, seems to be fine.

Reply:
The test on `config_large_gcn.yaml` shows that it takes 20 minutes per epoch, which is 10 minutes longer than before...

Reply:
Perhaps it's the `gradient_norm` that's taking a long time?

Reply:
Or perhaps the logic is wrong and things gets logged more often than once every 1000 steps?

Reply:
I'll look into that, currently I'm not logging any metrics only the losses. Are "preds" and "targets" only used for metric logging here?

BTW, I found a way to filter 0 losses from different replica or gradient accumulation. `oncatenated_metrics_logs[key] = concatenated_metrics_logs[key].mean()` is actually averaging on r`eplica * gradien_accu * device_iter`. In `ipu_wrapper.py`, there is another averaging `outputs["loss"] = outputs["loss"].mean()`. We might just need to filter out the 0s before doing the averaging. From another point of view, as the total loss is the sum of the 5 task losses, it would not end up with zero anyway I think? We may just need to filter on individual loss for the reporting purpose.

Reply:
> Or perhaps the logic is wrong and things gets logged more often than once every 1000 steps?

maybe it's not that bad after all, it's taking around 22 hours for 100 epochs, compared to 19 hours previously, it is quite ok I think.

Reply:
I removed the throughput computation and the `get_gradient_norm` if `self.skip_log_train_metrics is True`, so things should be a bit faster now